src/tools/database/elasticsearch.ts

```diff
@@ -0,0 +1,223 @@
+  connection: ClientOptions;
+}
+
+export type ElasticSearchToolResult =
+  | CatIndicesResponse
+  | IndicesGetMappingResponse
+  | SearchResponse;
+
+export const ElasticSearchAction = {
+  ListIndices: "LIST_INDICES",
+  getIndexDetails: "GET_INDEX_DETAILS",
+  search: "SEARCH",
+};
+
+export class ElasticSearchTool extends Tool<
+  JSONToolOutput<ElasticSearchToolResult>,
+  ElasticSearchToolOptions,
+  BaseToolRunOptions
+> {
+  name = "ElasticSearchTool";
+
+  description = `Can query data from an ElasticSearch database. IMPORTANT: strictly follow this order of actions:
+   1. ${ElasticSearchAction.ListIndices} - retrieve a list of available indices
+   2. ${ElasticSearchAction.getIndexDetails} - get details of index fields
+   3. ${ElasticSearchAction.search} - perform search or aggregation query on a specific index or pass the original user query without modifications if it's a valid JSON ElasticSearch query`;
+
+  inputSchema() {
+    return z.object({
+      action: z
+        .nativeEnum(ElasticSearchAction)
+        .describe(
+          "The action to perform. LIST_INDICES lists all indices, GET_INDEX_DETAILS fetches details for a specified index, and SEARCH executes a search or aggregation query",
+        ),
+      indexName: z
+        .string()
+        .optional()
+        .describe("The name of the index to query, required for GET_INDEX_DETAILS and SEARCH"),
+      query: z
+        .string()
+        .optional()
+        .describe("Valid ElasticSearch JSON search or aggregation query for SEARCH action"),
+      start: z.coerce
+        .number()
+        .int()
+        .min(0)
+        .default(0)
+        .optional()
+        .describe(
+          "The record index from which the query will start. Increase by the size of the query to get the next page of results",
+        ),
+      size: z.coerce
+        .number()
+        .int()
+        .min(0)
+        .max(10)
+        .default(10)
+        .optional()
+        .describe("How many records will be retrieved from the ElasticSearch query. Maximum is 10"),
+    });
+  }
+
+  protected validateInput(
+    schema: AnyToolSchemaLike,
+    input: unknown,
+  ): asserts input is ToolInput<this> {
+    super.validateInput(schema, input);
+    if (input.action === ElasticSearchAction.getIndexDetails && !input.indexName) {
+      throw new ToolInputValidationError(
+        `Index name is required for ${ElasticSearchAction.getIndexDetails} action.`,
+      );
+    }
+    if (input.action === ElasticSearchAction.search && (!input.indexName || !input.query)) {
+      throw new ToolInputValidationError(
+        `Both index name and query are required for ${ElasticSearchAction.search} action.`,
+      );
+    }
+    if (input.action === ElasticSearchAction.search && input.query) {
+      try {
+        JSON.parse(input.query);
+      } catch (error) {
+        throw new ToolInputValidationError(`Invalid JSON format for query ${error}`);
```

Comment:
Could you try if the formatting improves when you do this instead?

```
throw new ToolInputValidationError(`Invalid JSON format for query.`, [error]);
```

Reply:
Unfortunately, the agent is still not recovering:

Agent (tool_input) 🤖 :  {"action":"SEARCH","indexName":"kibana_sample_data_flights","query":"{\"query\":{\"bool\":{\"filter\":[{\"term\":{\"OriginCityName\":\"Cape Town\"}},{\"term\":{\"DestCityName\":\"Venice\"}}]},\"aggs\":{\"avg_ticket_price\":{\"avg\":{\"field\":\"AvgTicketPrice\"}}}}","start":0,"size":10}
Agent (tool_caption) 🤖 :  Searching for the average ticket price of all flights from Cape Town to Venice.
Agent (tool_output) 🤖 :  ToolInputValidationError: Invalid JSON format for query.

HINT: If you're convinced that the input was correct but the function cannot process it then use a different function or say I don't know.
Agent 🤖 :  [LinePrefixParserError: The generated output does not adhere to the schema.
Nothing valid has been parsed yet!] {
  isFatal: true,
  isRetryable: false,
  context: {
    lines: '',
    excludedLines: '',
    finalState: {},
    partialState: {}
  },
  [errors]: []
}
[AgentError: Error has occurred!] {
  isFatal: false,
  isRetryable: true,
  context: {},
  [errors]: [
    [LinePrefixParserError: The generated output does not adhere to the schema.
    Nothing valid has been parsed yet!] {
      isFatal: true,
      isRetryable: false,
      context: {
        lines: '',
        excludedLines: '',
        finalState: {},
        partialState: {}
      },
      [errors]: []
    }
  ]
}

Reply:
You can try to apply the `parseBrokenJson` function to fix the JSON generated by the LLM.